File: src/adapters/anthropic.ts
import { AgenticGate } from "../gate.js";

export interface AnthropicToolUseBlock {
  type: "tool_use";
  id: string;
  name: string;
  input: unknown;
}

export interface AnthropicContentBlock {
  type: string;
  text?: string;
  [key: string]: unknown;
}

export interface AnthropicMessageResponse {
  content: AnthropicContentBlock[];
}

export interface AdapterOutcome {
  /** true when the model returned a final answer with no tool_use blocks */
  done: boolean;
  /** the model's final text, only set when done is true */
  text?: string;
  /** messages to append to your conversation array, in order */
  messages: unknown[];
}

/**
 * Runs every tool_use block in an Anthropic Messages API response through the
 * gate and returns the messages to append: the assistant's own turn (content
 * replayed verbatim) plus one user turn carrying a tool_result block per call
 * — Anthropic requires all tool_results for one assistant turn to arrive
 * together in a single user message.
 */
export async function handleResponse(
  gate: AgenticGate,
  response: AnthropicMessageResponse
): Promise<AdapterOutcome> {
  const toolUseBlocks = response.content.filter((block) => block.type === "tool_use") as unknown as AnthropicToolUseBlock[];

  if (toolUseBlocks.length === 0) {
    const text = response.content.find((block) => block.type === "text")?.text ?? "";
    return { done: true, text, messages: [{ role: "assistant", content: response.content }] };
  }

  const toolResults: unknown[] = [];
  for (const block of toolUseBlocks) {
    const gateResult = await gate.interceptAndExecute(block.name, block.input);
    toolResults.push({
      type: "tool_result",
      tool_use_id: block.id,
      content: gateResult.success ? JSON.stringify(gateResult.data) : gateResult.error,
      ...(gateResult.success ? {} : { is_error: true }),
    });
  }

  return {
    done: false,
    messages: [
      { role: "assistant", content: response.content },
      { role: "user", content: toolResults },
    ],
  };
}

File: python/agentic_gate/adapters/anthropic.py
from __future__ import annotations

import json
from typing import Any

from ..gate import AgenticGate
from ._util import get, to_plain
from .types import AdapterOutcome


async def handle_response(gate: AgenticGate, response: Any) -> AdapterOutcome:
    """Runs every tool_use block in an Anthropic Messages API response
    through the gate and returns the messages to append: the assistant's own
    turn (content replayed verbatim) plus one user turn carrying a
    tool_result block per call — Anthropic requires all tool_results for one
    assistant turn to arrive together in a single user message.

    Accepts either the raw `anthropic` SDK response object or an equivalent
    plain dict."""
    content = get(response, "content")
    tool_use_blocks = [block for block in content if get(block, "type") == "tool_use"]

    if not tool_use_blocks:
        text = next((get(block, "text") for block in content if get(block, "type") == "text"), "")
        return AdapterOutcome(done=True, text=text or "", messages=[{"role": "assistant", "content": [to_plain(b) for b in content]}])

    tool_results = []
    for block in tool_use_blocks:
        name = get(block, "name")
        tool_input = get(block, "input")
        gate_result = await gate.intercept_and_execute(name, tool_input)

        result_block = {
            "type": "tool_result",
            "tool_use_id": get(block, "id"),
            "content": json.dumps(gate_result.data) if gate_result.success else gate_result.error,
        }
        if not gate_result.success:
            result_block["is_error"] = True
        tool_results.append(result_block)

    return AdapterOutcome(
        done=False,
        messages=[
            {"role": "assistant", "content": [to_plain(b) for b in content]},
            {"role": "user", "content": tool_results},
        ],
    )
